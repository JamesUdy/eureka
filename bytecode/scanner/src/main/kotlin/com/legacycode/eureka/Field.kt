package com.legacycode.eureka

data class Field(
  override val name: String,
  val descriptor: FieldDescriptor,
  override val owner: QualifiedType,
) : Member {
  override val signature: Signature
    get() = FieldSignature(name, QualifiedType(descriptor.type))
}
