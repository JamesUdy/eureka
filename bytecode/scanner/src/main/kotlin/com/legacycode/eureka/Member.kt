package com.legacycode.eureka

sealed interface Member {
  val name: String
  val signature: Signature
  val owner: QualifiedType
}
