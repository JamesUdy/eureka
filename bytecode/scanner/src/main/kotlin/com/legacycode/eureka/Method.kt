package com.legacycode.eureka

data class Method(
  override val name: String,
  val descriptor: MethodDescriptor,
  override val owner: QualifiedType,
) : Member {
  val isLambda: Boolean
    get() = name.startsWith("lambda\$") || /* Java */
      name.contains("\$lambda-") || /* Kotlin 1.7.10 */
      name.contains("\$lambda\$") /* Kotlin 1.7.20 */

  val isBridge: Boolean
    get() = name.startsWith("access\$") /* Kotlin */

  val isCopyConstructor: Boolean
    get() = name.startsWith("copy\$default") /* Kotlin */

  val isSynthetic: Boolean
    get() = isLambda || isBridge || isCopyConstructor

  override val signature: Signature
    get() = MethodSignature.from(
      name,
      descriptor.parameters,
      descriptor.returnType,
    )
}
