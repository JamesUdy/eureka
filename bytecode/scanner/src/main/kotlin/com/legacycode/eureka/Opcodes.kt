package com.legacycode.eureka

object Opcodes {
  const val iconst_m1 = 0x02
  const val iconst_0 = 0x03
  const val iconst_1 = 0x04
  const val iconst_2 = 0x05
  const val iconst_3 = 0x06
  const val iconst_4 = 0x07
  const val iconst_5 = 0x08
  const val areturn = 0xb0

  const val invokespecial = 0xb7

  fun isIntInsn(opcode: Int): Boolean {
    return opcode == iconst_m1 ||
      opcode == iconst_0 ||
      opcode == iconst_1 ||
      opcode == iconst_2 ||
      opcode == iconst_3 ||
      opcode == iconst_4 ||
      opcode == iconst_5
  }
}
