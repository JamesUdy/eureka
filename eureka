#!/bin/bash

set -e

./gradlew :cli:installDist -q
./cli/build/install/eureka/bin/eureka $*
