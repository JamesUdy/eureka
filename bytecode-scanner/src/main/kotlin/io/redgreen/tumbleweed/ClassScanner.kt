package io.redgreen.tumbleweed

import java.io.BufferedInputStream
import java.io.File
import net.bytebuddy.jar.asm.ClassReader
import net.bytebuddy.jar.asm.ClassVisitor
import net.bytebuddy.jar.asm.FieldVisitor
import net.bytebuddy.jar.asm.MethodVisitor
import net.bytebuddy.jar.asm.Opcodes.ASM9
import org.slf4j.Logger
import org.slf4j.LoggerFactory

const val ASM_API_VERSION = ASM9

object ClassScanner {
  private val logger: Logger = LoggerFactory.getLogger(ClassScanner::class.java)

  fun scan(classFile: File): ClassStructure {
    val outClassFilesQueue = ArrayDeque(listOf(classFile))
    val visitedClassFiles = mutableSetOf<File>()
    val classStructures = mutableListOf<ClassStructure>()

    while (outClassFilesQueue.isNotEmpty()) {
      val classFileToScan = outClassFilesQueue.removeFirst()
      if (classFileToScan !in visitedClassFiles) {
        visitedClassFiles.add(classFileToScan)

        logger.debug("Scanning class file: {}", classFileToScan.absolutePath)
        val classStructure = classStructure(classFileToScan, outClassFilesQueue, visitedClassFiles)
        classStructures.add(classStructure)
      }
    }

    val simplify = classStructures.combine().first().simplify()
    return simplify
  }

  private fun classStructure(
    classFile: File,
    outClassFilesQueue: ArrayDeque<File>,
    visitedClassFiles: Set<File>,
  ): ClassStructure {
    val outFields = mutableListOf<Field>()
    val outMethods = mutableListOf<Method>()
    val outRelationships = mutableListOf<Relationship>()

    val classInfo = ClassInfo.from(classFile)
    val topLevelType = classInfo.topLevelType

    val classVisitor = object : ClassVisitor(ASM_API_VERSION) {
      override fun visitField(
        access: Int,
        name: String?,
        descriptor: String?,
        signature: String?,
        value: Any?,
      ): FieldVisitor {
        logger.debug("Visiting field: {}", name)

        outFields.add(Field(name!!, FieldDescriptor.from(descriptor!!), topLevelType))
        return object : FieldVisitor(ASM_API_VERSION) { /* no-op */ }
      }

      override fun visitMethod(
        access: Int,
        name: String?,
        descriptor: String?,
        signature: String?,
        exceptions: Array<out String>?,
      ): MethodVisitor {
        logger.debug("Visiting method: {}", name)

        val method = Method(name!!, MethodDescriptor(descriptor!!), topLevelType)
        outMethods.add(method)
        return InstructionScanner.scan(topLevelType, method, outRelationships)
      }

      override fun visitInnerClass(name: String, outerName: String?, innerName: String?, access: Int) {
        logger.debug("Visiting inner class: {}", name)

        val innerClassFile = classFile.parentFile.resolve("${name.substring(name.lastIndexOf('/') + 1)}.class")
        if (innerClassFile.exists() && innerClassFile !in visitedClassFiles) {
          logger.debug("Adding new file to scan: ${innerClassFile.canonicalPath}")
          outClassFilesQueue.addLast(innerClassFile)
        }
      }
    }

    ClassReader(BufferedInputStream(classFile.inputStream())).accept(classVisitor, 0)

    val allMembersInRelationships = outRelationships.flatMap { listOf(it.source, it.target) }
    val methodsFromRelationships = allMembersInRelationships
      .filterIsInstance<Method>()
      .filter { it.owner == topLevelType }
      .distinct()
    val fieldsFromRelationships = allMembersInRelationships.filterIsInstance<Field>().distinct()

    val missingFields = fieldsFromRelationships - outFields.toSet()
    val missingMethods = methodsFromRelationships - outMethods.toSet()

    return ClassStructure(
      classInfo.packageName,
      classInfo.className,
      outFields + missingFields,
      outMethods + missingMethods,
      outRelationships.toList(),
    )
  }

  private fun List<ClassStructure>.combine(): List<ClassStructure> {
    val combinedClassStructures = mutableListOf<ClassStructure>()

    for (classStructure in this) {
      val syntheticInnerClassRelationships = classStructure.relationships
        .filter { it.type == Relationship.Type.Calls }
        .filter { (it.target as Method).name == "<init>" }
        .filter { it.target.owner != classStructure.className }

      val referencesRelationships = mutableListOf<Relationship>()
      for (relationship in syntheticInnerClassRelationships) {
        val innerClassConstructor = relationship.target as Method
        val innerClass = this.find { innerClassConstructor.owner.endsWith(it.className) }!!

        val outerClassAccessorsRelationships = innerClass.relationships
          .filter { it.target.owner.endsWith(classStructure.className) }

        for (accessorRelationship in outerClassAccessorsRelationships) {
          referencesRelationships.add(
            Relationship(relationship.source, accessorRelationship.target, Relationship.Type.References)
          )
        }
      }
      val updatedClassStructure = classStructure
        .copy(relationships = classStructure.relationships + referencesRelationships)
      combinedClassStructures.add(updatedClassStructure)
    }

    return combinedClassStructures.toList()
  }
}
