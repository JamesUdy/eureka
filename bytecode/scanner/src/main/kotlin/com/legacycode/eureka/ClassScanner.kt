package com.legacycode.eureka

import java.io.BufferedInputStream
import java.io.File
import net.bytebuddy.jar.asm.ClassReader
import net.bytebuddy.jar.asm.ClassVisitor
import net.bytebuddy.jar.asm.FieldVisitor
import net.bytebuddy.jar.asm.Handle
import net.bytebuddy.jar.asm.Label
import net.bytebuddy.jar.asm.MethodVisitor
import net.bytebuddy.jar.asm.Opcodes.ACC_STATIC
import net.bytebuddy.jar.asm.Opcodes.ASM9
import org.slf4j.Logger
import org.slf4j.LoggerFactory

typealias ConstantPool = MutableMap<Any?, Field>

const val ASM_API_VERSION = ASM9

fun constantPool(): ConstantPool = mutableMapOf()

object ClassScanner {
  private val logger: Logger = LoggerFactory.getLogger(ClassScanner::class.java)

  fun scan(classFile: File): ClassStructure {
    val outClassFilesQueue = ArrayDeque(listOf(classFile))
    val visitedClassFiles = mutableSetOf<File>()
    val classStructures = mutableListOf<ClassStructure>()

    while (outClassFilesQueue.isNotEmpty()) {
      val classFileToScan = outClassFilesQueue.removeFirst()
      if (classFileToScan !in visitedClassFiles) {
        visitedClassFiles.add(classFileToScan)

        logger.debug("Scanning class file: {}", classFileToScan.absolutePath)
        val classStructure = classStructure(classFileToScan, visitedClassFiles, outClassFilesQueue)
        classStructures.add(classStructure)
      }
    }

    logger.debug("Condensing class structures.")
    val classStructure = classStructures.combine()

    logger.debug("Simplifying class structure.")
    return classStructure.normalize()
  }

  private fun classStructure(
    classFile: File,
    visitedClassFiles: Set<File>,
    outClassFilesQueue: ArrayDeque<File>,
  ): ClassStructure {
    lateinit var superClassName: String
    val outInterfaces = mutableListOf<QualifiedType>()
    val outFields = mutableListOf<Field>()
    val outMethods = mutableListOf<Method>()
    val outRelationships = mutableListOf<Relationship>()

    val classInfo = ClassInfo.from(classFile)
    val topLevelType = classInfo.type

    val constantPool = constantPool()

    val classVisitor = object : ClassVisitor(ASM_API_VERSION) {
      override fun visit(
        version: Int,
        access: Int,
        name: String?,
        signature: String?,
        superName: String,
        interfaces: Array<out String>?,
      ) {
        super.visit(version, access, name, signature, superName, interfaces)
        superClassName = superName.replace('/', '.')
        interfaces?.forEach { interfaceName ->
          outInterfaces.add(QualifiedType(interfaceName.replace('/', '.')))
        }
      }

      override fun visitField(
        access: Int,
        name: String?,
        descriptor: String?,
        signature: String?,
        value: Any?,
      ): FieldVisitor {
        logger.debug("Visiting field: {}", name)

        val field = Field(name!!, FieldDescriptor.from(descriptor!!), topLevelType)
        if (access and ACC_STATIC != 0) {
          constantPool[value] = field
          logger.debug("Adding to constant pool: {} = {}", name, value)
        }

        outFields.add(field)
        return object : FieldVisitor(ASM_API_VERSION) { /* no-op */ }
      }

      override fun visitMethod(
        access: Int,
        name: String?,
        descriptor: String?,
        signature: String?,
        exceptions: Array<out String>?,
      ): MethodVisitor {
        logger.debug("Visiting method: {}", name)

        val method = Method(name!!, MethodDescriptor(descriptor!!), topLevelType)
        outMethods.add(method)

        return object : MethodVisitor(ASM_API_VERSION) {
          private var maybeConstantFieldReferencedByInsn: Field? = null

          override fun visitFieldInsn(
            opcode: Int,
            owner: String,
            fieldName: String,
            fieldDescriptor: String,
          ) {
            logger.debug("Visiting field instruction ({}): {}/{}", opcode.instruction, owner, fieldName)

            val field = Field(fieldName, FieldDescriptor.from(fieldDescriptor), QualifiedType(owner))
            val relationship = Relationship(method, field, Relationship.Type.from(opcode))
            if (owner == topLevelType.name) {
              logger.debug("Adding relationship: {}", relationship)
              outRelationships.add(relationship)
            } else {
              logger.debug("Skipping relationship: {}", relationship)
            }
          }

          override fun visitMethodInsn(
            opcode: Int,
            owner: String,
            methodName: String,
            methodDescriptor: String,
            isInterface: Boolean,
          ) {
            logger.debug("Visiting method instruction ({}): {}/{}", opcode.instruction, owner, methodName)

            if (topLevelType.name.startsWith(owner) || owner.startsWith("${topLevelType.name}\$")) {
              val callee = Method(methodName, MethodDescriptor(methodDescriptor), QualifiedType(owner))
              val relationship = Relationship(method, callee, Relationship.Type.from(opcode))

              if (method.name == "<clinit>" && opcode == Opcodes.invokespecial) {
                logger.debug("Skipping synthetic bridge call from static block: {}", relationship)
              } else {
                logger.debug("Adding relationship: {}", relationship)
                outRelationships.add(relationship)
              }
            } else {
              logger.debug("Skipping method instruction ({}): {}/{}", opcode.instruction, owner, methodName)
            }

            if (maybeConstantFieldReferencedByInsn != null) {
              val relationship = Relationship(method, maybeConstantFieldReferencedByInsn!!, Relationship.Type.Reads)

              logger.debug("Adding relationship: {}", relationship)
              outRelationships.add(relationship)
              maybeConstantFieldReferencedByInsn = null
            }
          }

          override fun visitInvokeDynamicInsn(
            name: String?,
            descriptor: String?,
            bootstrapMethodHandle: Handle?,
            vararg bootstrapMethodArguments: Any?,
          ) {
            logger.debug(
              "Visiting invoke dynamic instruction: {}, {}",
              name,
              bootstrapMethodArguments.contentToString()
            )

            if (bootstrapMethodArguments.size > 1) {
              val methodHandle = bootstrapMethodArguments[1] as Handle
              val callee = Method(methodHandle.name, MethodDescriptor(methodHandle.desc), topLevelType)
              outRelationships.add(Relationship(method, callee, Relationship.Type.Calls))
            } else {
              logger.debug("Ignoring dynamic instruction: {}{}", name, descriptor)
            }
            super.visitInvokeDynamicInsn(name, descriptor, bootstrapMethodHandle, *bootstrapMethodArguments)
          }

          override fun visitLdcInsn(value: Any?) {
            logger.debug("Visiting ldc instruction: {}", value)

            maybeConstantFieldReferencedByInsn = constantPool[value]
            super.visitLdcInsn(value)
          }

          override fun visitIntInsn(opcode: Int, operand: Int) {
            logger.debug("Visiting int instruction: {}", opcode.instruction)

            maybeConstantFieldReferencedByInsn = constantPool[operand]
            super.visitIntInsn(opcode, operand)
          }

          override fun visitInsn(opcode: Int) {
            logger.debug("Visiting instruction: {}", opcode.instruction)

            if (opcode == Opcodes.iconst_m1) {
              maybeConstantFieldReferencedByInsn = constantPool[-1]
            } else if (Opcodes.isIntInsn(opcode)) {
              maybeConstantFieldReferencedByInsn = constantPool[opcode - Opcodes.iconst_0]
            }

            if (opcode == Opcodes.areturn && maybeConstantFieldReferencedByInsn != null) {
              val relationship = Relationship(method, maybeConstantFieldReferencedByInsn!!, Relationship.Type.Reads)
              logger.debug("Adding relationship: {}", relationship)
              outRelationships.add(relationship)
              maybeConstantFieldReferencedByInsn = null
            }
            super.visitInsn(opcode)
          }

          override fun visitJumpInsn(opcode: Int, label: Label?) {
            logger.debug("Visiting jump instruction: {}", opcode.instruction)

            if (maybeConstantFieldReferencedByInsn != null) {
              val relationship = Relationship(method, maybeConstantFieldReferencedByInsn!!, Relationship.Type.Reads)

              logger.debug("Adding relationship: {}", relationship)
              outRelationships.add(relationship)
              maybeConstantFieldReferencedByInsn = null
            }
            super.visitJumpInsn(opcode, label)
          }
        }
      }

      override fun visitInnerClass(name: String, outerName: String?, innerName: String?, access: Int) {
        logger.debug("Visiting inner class: {}", name)
        val innerClassFilename = "${name.substring(name.lastIndexOf('/') + 1)}.class"

        val innerClassFile = classFile.parentFile.resolve(innerClassFilename)
        if (innerClassFile.exists() && innerClassFile !in visitedClassFiles) {
          logger.debug("Adding new file to scan: ${innerClassFile.canonicalPath}")
          outClassFilesQueue.addLast(innerClassFile)
        }
      }
    }

    ClassReader(BufferedInputStream(classFile.inputStream())).accept(classVisitor, 0)

    return ClassStructure(
      classInfo.packageName,
      classInfo.className,
      QualifiedType(superClassName),
      outInterfaces.toList(),
      outFields,
      outMethods,
      outRelationships.toList(),
      topLevelType,
    )
  }
}

@Suppress("MagicNumber")
internal val Int.instruction: String
  get() {
    return when (this) {
      Opcodes.`return` -> "return"
      Opcodes.getstatic -> "getstatic"
      Opcodes.putstatic -> "putstatic"
      Opcodes.putfield -> "putfield"
      Opcodes.getfield -> "getfield"
      Opcodes.invokespecial -> "invokespecial"
      Opcodes.invokevirtual -> "invokevirtual"
      Opcodes.invokestatic -> "invokestatic"
      Opcodes.invokeinterface -> "invokeinterface"
      Opcodes.bipush -> "bipush"
      Opcodes.iconst_m1 -> "iconst_m1"
      Opcodes.iconst_0 -> "iconst_0"
      Opcodes.iconst_1 -> "iconst_1"
      Opcodes.iconst_2 -> "iconst_2"
      Opcodes.iconst_3 -> "iconst_3"
      Opcodes.iconst_4 -> "iconst_4"
      Opcodes.iconst_5 -> "iconst_5"
      Opcodes.if_icmpne -> "if_icmpne"
      Opcodes.areturn -> "areturn"
      else -> "unmapped (${"0x%02x".format(this)})})"
    }
  }
