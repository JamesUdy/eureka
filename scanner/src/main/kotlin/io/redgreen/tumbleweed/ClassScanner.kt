package io.redgreen.tumbleweed

import java.io.BufferedInputStream
import net.bytebuddy.jar.asm.ClassReader
import net.bytebuddy.jar.asm.ClassVisitor
import net.bytebuddy.jar.asm.FieldVisitor
import net.bytebuddy.jar.asm.MethodVisitor
import net.bytebuddy.jar.asm.Opcodes.ASM7

object ClassScanner {
  fun scan(location: ClassFileLocation): ClassStructure {
    var className: String? = null
    var packageName: String? = null
    val outFields = mutableListOf<Field>()
    val outMethods = mutableListOf<Method>()

    val classVisitor = object : ClassVisitor(ASM7) {
      override fun visit(
        version: Int,
        access: Int,
        name: String,
        signature: String?,
        superName: String?,
        interfaces: Array<out String>?,
      ) {
        name.split("/").let { fqClassNameParts ->
          packageName = fqClassNameParts.dropLast(1).joinToString(".")
          className = fqClassNameParts.last()
        }
      }

      override fun visitField(
        access: Int,
        name: String?,
        descriptor: String?,
        signature: String?,
        value: Any?,
      ): FieldVisitor {
        return FieldScanner.scan(name, descriptor, outFields)
      }

      override fun visitMethod(
        access: Int,
        name: String?,
        descriptor: String?,
        signature: String?,
        exceptions: Array<out String>?
      ): MethodVisitor {
        return MethodScanner.scan(name, descriptor, outMethods)
      }
    }

    ClassReader(BufferedInputStream(location.file.inputStream())).accept(classVisitor, 0)

    return ClassStructure(packageName!!, className!!, outFields.toList(), outMethods.toList())
  }
}
