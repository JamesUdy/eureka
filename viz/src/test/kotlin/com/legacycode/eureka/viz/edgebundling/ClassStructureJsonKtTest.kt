package com.legacycode.eureka.viz.edgebundling

import com.legacycode.eureka.ClassScanner
import com.legacycode.eureka.samples.AccessSuperClassMembers
import com.legacycode.eureka.samples.InterfaceImplementation
import com.legacycode.eureka.samples.OnlyMethods
import com.legacycode.eureka.testing.SampleClass
import org.approvaltests.JsonApprovals
import org.junit.jupiter.api.Test

internal class ClassStructureJsonKtTest {
  private val scanner = ClassScanner()

  @Test
  internal fun `serialize class structure with a super type as edge bundling graph JSON`() {
    // given
    val classFile = SampleClass.Kotlin(AccessSuperClassMembers::class).file
    val classStructure = scanner.scan(classFile)

    // when & then
    JsonApprovals.verifyAsJson(classStructure.toGraph())
  }

  @Test
  internal fun `serialize class structure that does not inherit a super type`() {
    // given
    val classFile = SampleClass.Kotlin(OnlyMethods::class).file
    val classStructure = scanner.scan(classFile)

    // when & then
    JsonApprovals.verifyAsJson(classStructure.toGraph())
  }

  @Test
  internal fun `serialize class structure that implements an interface`() {
    // given
    val classFile = SampleClass.Kotlin(InterfaceImplementation::class).file
    val classStructure = scanner.scan(classFile)

    // when & then
    JsonApprovals.verifyAsJson(classStructure.toGraph())
  }
}
