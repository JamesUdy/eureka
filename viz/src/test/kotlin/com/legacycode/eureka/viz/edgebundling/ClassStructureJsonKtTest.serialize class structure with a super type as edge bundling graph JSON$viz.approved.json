{
  "nodes": [
    {
      "id": "void \u003cinit\u003e()",
      "group": 2
    },
    {
      "id": "void bar()",
      "group": 2
    },
    {
      "id": "void baz()",
      "group": 2
    },
    {
      "id": "void foo()",
      "group": 2
    },
    {
      "id": "String getBaz()",
      "group": 2
    }
  ],
  "links": [
    {
      "source": "void bar()",
      "target": "void foo()",
      "value": 1
    },
    {
      "source": "void baz()",
      "target": "String getBaz()",
      "value": 1
    }
  ],
  "meta": {
    "classInfo": {
      "name": "com.legacycode.eureka.samples.AccessSuperClassMembers",
      "extends": "com.legacycode.eureka.samples.SuperClass"
    }
  }
}
